grails-app/services/org/pih/warehouse/inventory/StockMovementService.groovy

```diff
@@ -2813,4 +2815,34 @@ class StockMovementService {
         }
         return true
     }
+
+    void rollbackApproval(String stockMovementId) {
+        StockMovement stockMovement = getStockMovement(stockMovementId)
+        // TODO: remove .get() from current user in Grails 3 (with .get it will return null)
+        if (!canRollbackApproval(AuthService.currentUser.get(), stockMovement)) {
+            String errorMessage = applicationTagLib.message(
+                    code: "request.errors.approval.rollback.message",
+                    default: "Unable to rollback approval due to insufficient permissions",
+            )
+            throw new IllegalAccessException(errorMessage)
+        }
+
+        Requisition requisition = stockMovement?.requisition
+        if (requisition.status in [RequisitionStatus.APPROVED, RequisitionStatus.REJECTED]) {
+            requisitionService.rollbackLastEvent(requisition)
+            requisition.status = RequisitionStatus.PENDING_APPROVAL
+            requisition.approvedBy = null
+            requisition.rejectedBy = null
+            requisition.dateApproved = null
+            requisition.dateRejected = null
```

Comment:
I don't know if we want to remove the rejection, do we? If a user has made a request and it has been rejected and then approved, don't we want the rollback to only undo the approval? Or does approval also remove the rejected?

Reply:
A request can't be rejected and then approved. It has to be rolled back before approval so it won't have approved and rejected data.